### PR TITLE
drivers: modem: add Quectel EG25-G

### PR DESCRIPTION
**Pull request by ldenefle**

### CODE REVIEW

**Comment by ycsin** [comment]
Commit should have message and sign-off

**Comment by bbilas** [comment]
Would be not better if we put all of those definitions into separate headers? I mean something like

<pre>
...snip...
#if DT_HAS_COMPAT_STATUS_OKAY(quectel_eg25_g)
#include "quectel_eg25_g.h"
#endif

#if DT_HAS_COMPAT_STATUS_OKAY(quectel_egXX)
#include "quectel_egXX.h"
#endif
...snip...
</pre>

(or use some macrology)
where `quectel_eg25_g.h` would contain:

<pre>
MODEM_CHAT_SCRIPT_CMDS_DEFINE(quectel_eg25_g_init_chat_script_cmds,
			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CFUN=4", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CMEE=1", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG=1", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGREG=1", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CEREG=1", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG?", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CEREG?", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGREG?", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGSN", imei_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGMM", cgmm_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT+CMUX=0,0,5,127,10,3,30,10,2",
							      0));

MODEM_CHAT_SCRIPT_DEFINE(quectel_eg25_g_init_chat_script, quectel_eg25_g_init_chat_script_cmds,
			 abort_matches, modem_cellular_chat_callback_handler, 10);

MODEM_CHAT_SCRIPT_CMDS_DEFINE(quectel_eg25_g_dial_chat_script_cmds,
			      MODEM_CHAT_SCRIPT_CMD_RESP_MULT("AT+CGACT=0,1", allow_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGDCONT=1,\"IP\","
							 "\""CONFIG_MODEM_CELLULAR_APN"\"",
							 ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CFUN=1", ok_match),
			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("ATD*99***1#", 0),);

MODEM_CHAT_SCRIPT_DEFINE(quectel_eg25_g_dial_chat_script, quectel_eg25_g_dial_chat_script_cmds,
			 dial_abort_matches, modem_cellular_chat_callback_handler, 10);
</pre>

**Comment by bjarki-trackunit** [comment]
@bbilas We definitely need to separate it out somehow, I still don't know exactly what the best solution is though, and it should be its own PR I think :)

**Review by bjarki-andreasen — DISMISSED** [review]

File: dts/bindings/modem/quectel,eg25-g.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+properties:
+  mdm-reset-gpios:
+    type: phandle-array
+
+  mdm-dtr-gpios:
+    type: phandle-array
+
+  mdm-wdisable-gpios:
+    type: phandle-array
```
Comment (rerickson1) [inline dts/bindings/modem/quectel,eg25-g.yaml:15]: Do you want any of these to be required?

Comment (ldenefle) [reply dts/bindings/modem/quectel,eg25-g.yaml]: yes good idea, although the current cellular driver ignores dtr and wdisable

Comment (rerickson1) [reply dts/bindings/modem/quectel,eg25-g.yaml]: Maybe you want those two optional then.

**Review by rerickson1 — DISMISSED** [review]

**Review by ycsin — DISMISSED** [review]

**Review by bjarki-andreasen — DISMISSED** [review]

**Comment by ldenefle** [comment]
rebased and added periodic chat scripts

**Review by ycsin — COMMENTED** [review]
Is it possible to generalize Quectel chat scripts? They look pretty similar, [EC21](https://github.com/ycsin/zephyr/blob/develop_v2.7-1/drivers/modem/gsm_ppp.c#L1019) and its variants would be using the same init process as well

**Comment by ldenefle** [comment]
They all look very similar tbf, I think there's a potential for a refacto there but should it be done in a separate PR ?
I also don't have other quectel modems with me so won't be able to test any changes.

**Comment by ldenefle** [comment]
Hello, what's the update on the status of this PR ?
I could potentially merge the quectel init scripts but I could also do this in a separate PR, I'd like to get this one merged :)

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by bjarki-trackunit** [comment]
Sorry, forgot about it :) @rerickson1  @ycsin

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**